internal/controller/pod_controller_test.go

```diff
@@ -0,0 +1,265 @@
+			for _, pod := range pods.Items {
+				if !existsSchedulingGate(&pod) {
+					numSchedulable += 1
+				}
+			}
+			g.Expect(numSchedulable).To(Equal(8))
+		}).Should(Succeed())
+	})
+
+	It("should remove scheduling gate when annotation force-removed", func() {
+		testName := "remove-scheduling-gate-fail-safe"
+		namespace := &corev1.Namespace{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: testName,
+			},
+		}
+		err := k8sClient.Create(ctx, namespace)
+		Expect(err).NotTo(HaveOccurred())
+
+		createNewPod(0, testName)
+
+		pod := &corev1.Pod{}
+		Eventually(func(g Gomega) {
+			err = k8sClient.Get(ctx, client.ObjectKey{Name: fmt.Sprintf("%s-pod-%d", testName, 0), Namespace: testName}, pod)
+			g.Expect(err).NotTo(HaveOccurred())
+			g.Expect(pod.Spec.SchedulingGates).NotTo(ConsistOf(corev1.PodSchedulingGate{Name: constants.PodSchedulingGateName}))
+		}).Should(Succeed())
+
+		pod = createNewPod(1, testName)
+		delete(pod.Annotations, constants.CatGateImagesHashAnnotation)
+		err = k8sClient.Update(ctx, pod)
+		Expect(err).NotTo(HaveOccurred())
+
+		Eventually(func(g Gomega) {
+			err = k8sClient.Get(ctx, client.ObjectKey{Name: fmt.Sprintf("%s-pod-%d", testName, 1), Namespace: testName}, pod)
+			g.Expect(err).NotTo(HaveOccurred())
+			g.Expect(pod.Spec.SchedulingGates).NotTo(ConsistOf(corev1.PodSchedulingGate{Name: constants.PodSchedulingGateName}))
+		}).Should(Succeed())
+	})
+
+	It("should limit number of schedulable pods based on status", func() {
+		testName := "multiple-pods-with-some-running-pods"
+		namespace := &corev1.Namespace{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: testName,
+			},
+		}
+		err := k8sClient.Create(ctx, namespace)
+		Expect(err).NotTo(HaveOccurred())
+
+		for i := 0; i < 8; i++ {
+			createNewPod(i, testName)
+		}
+		pods := &corev1.PodList{}
+
+		Eventually(func(g Gomega) {
+			err := k8sClient.List(ctx, pods, &client.ListOptions{Namespace: testName})
+			g.Expect(err).NotTo(HaveOccurred())
+			numSchedulable := 0
+			for _, pod := range pods.Items {
+				if !existsSchedulingGate(&pod) {
+					numSchedulable += 1
+				}
+			}
+			g.Expect(numSchedulable).To(Equal(1))
+		}).Should(Succeed())
+		updateStatusForPodWithSchedulingGate(testName)
+
+		Eventually(func(g Gomega) {
+			err := k8sClient.List(ctx, pods, &client.ListOptions{Namespace: testName})
+			g.Expect(err).NotTo(HaveOccurred())
+			numSchedulable := 0
+			for _, pod := range pods.Items {
+				if !existsSchedulingGate(&pod) {
+					numSchedulable += 1
+				}
+			}
+			g.Expect(numSchedulable).To(Equal(3))
+		}).Should(Succeed())
+		updateStatusForOnePodWithSchedulingGate(testName)
+
+		Eventually(func(g Gomega) {
+			err := k8sClient.List(ctx, pods, &client.ListOptions{Namespace: testName})
+			g.Expect(err).NotTo(HaveOccurred())
+			numSchedulable := 0
+			for _, pod := range pods.Items {
+				if !existsSchedulingGate(&pod) {
+					numSchedulable += 1
+				}
+			}
+			g.Expect(numSchedulable).To(Equal(6))
+		}).Should(Succeed())
+	})
+})
```

Comment:
It is desirable to add a test where one node has multiple running pods and only one additional pod would be scheduled.